src/DependencyInjection/DI/src/CallSiteFormatter.cs

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Text;
+using Microsoft.Extensions.DependencyInjection.ServiceLookup;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    internal class CallSiteFormatter: CallSiteVisitor<CallSiteFormatter.CallSiteFormatterContext, object>
+    {
+        internal static CallSiteFormatter Instance = new CallSiteFormatter();
+
+        private CallSiteFormatter()
+        {
+        }
+
+        public string Format(ServiceCallSite callSite)
+        {
+            var stringBuilder = new StringBuilder();
+
+            VisitCallSite(callSite, new CallSiteFormatterContext(stringBuilder, 0));
+
+            return stringBuilder.ToString();
+        }
+
+        protected override object VisitConstructor(ConstructorCallSite constructorCallSite, CallSiteFormatterContext argument)
+        {
+            argument.ContinueLine("new {0}", constructorCallSite.ImplementationType);
+
+            var newContext = argument.IncrementOffset();
+            foreach (var parameter in constructorCallSite.ParameterCallSites)
+            {
+                VisitCallSite(parameter, newContext);
+            }
+
+            return null;
+        }
+
+        protected override object VisitConstant(ConstantCallSite constantCallSite, CallSiteFormatterContext argument)
+        {
+            argument.ContinueLine("{0}", constantCallSite.DefaultValue ?? "<null>");
+
+            return null;
+        }
+
+        protected override object VisitServiceProvider(ServiceProviderCallSite serviceProviderCallSite, CallSiteFormatterContext argument)
+        {
+            argument.ContinueLine("<ServiceProvider>");
+
+            return null;
+        }
+
+        protected override object VisitServiceScopeFactory(ServiceScopeFactoryCallSite serviceScopeFactoryCallSite, CallSiteFormatterContext argument)
+        {
+            argument.ContinueLine("<ScopeFactory>");
+
+            return null;
+        }
+
+        protected override object VisitIEnumerable(IEnumerableCallSite enumerableCallSite, CallSiteFormatterContext argument)
+        {
+            argument.ContinueLine("IEnumerable<{0}> (size {1})", enumerableCallSite.ItemType, enumerableCallSite.ServiceCallSites.Length);
+
+            var newContext = argument.IncrementOffset();
+            foreach (var item in enumerableCallSite.ServiceCallSites)
+            {
+                VisitCallSite(item, newContext);
+            }
+
+            return null;
+        }
+
+        protected override object VisitFactory(FactoryCallSite factoryCallSite, CallSiteFormatterContext argument)
+        {
+            argument.ContinueLine("Factory ({0})", factoryCallSite.Factory.Method);
```

Comment:
The other "special token" values use `<>`, I think that might be appropriate here: `<Factory: [MethodName]>`